### PR TITLE
Azure security center

### PR DESCRIPTION
**Pull request by jochman**
fixed azure oauth2 with subscriptions

### CODE REVIEW

File: web/azureschandlers.go [hunk 1]
```diff
@@ -1,48 +1,57 @@
...
-	consent := r.FormValue("admin_consent")
-	tenant := r.FormValue("tenant")
-
-	if state == "" || tenant == "" || !strings.EqualFold(consent, "true") {
+	subid := r.FormValue("state")
+	code := r.FormValue("code")
+	if subid == "" || code == "" {
 		WriteError(w, &Error{"oauth_err", 401, "Missing consent", "Please make sure to accept the app request"})
-		logrus.Warnf("Missing something in the request - state: [%s], consent: [%s], tenant: [%s]", state, consent, tenant)
+		logrus.Warnf("Missing something in the request - code: [%s], subid: [%s], tenant: [%s]", subid, code)
```
Comment (liorblob) [inline web/azureschandlers.go]: Looks like you've put it backwards

File: web/azureschandlers.go [hunk 2]
```diff
@@ -1,48 +1,57 @@
...
 
 	if err == nil {
 		ac.r.RemoveToken(customer)
 	}
 
 	uu, err := uuid.NewRandom()
+
 	if err != nil {
-		logrus.Warnf("Could not create token - state: [%s], consent: [%s], tenant: [%s]", state, consent, tenant)
+		logrus.Warnf("Could not create token - subid: [%s], code: [%s], tenant: [%s]", subid, code, tenant)
```
Comment (liorblob) [inline web/azureschandlers.go]: Again...

File: web/azureschandlers.go [hunk 3]
```diff
@@ -52,22 +61,81 @@ func (ac *AppContext) loginOAuthAzureSC(w http.ResponseWriter, r *http.Request)
 
 	http.Redirect(w, r, fmt.Sprintf("/azuresc-details?t=%s&u=%s", tenant, id), http.StatusFound)
 }
+func getTenant(subid string) string {
+	SubUrl := fmt.Sprintf("https://management.azure.com/subscriptions/%s?api-version=2015-01-01", subid)
+	resp, _ := http.Get(SubUrl)
+
+	defer resp.Body.Close()
+	www := resp.Header.Get("WWW-Authenticate")
+	logrus.Warnf(www)
```
Comment (liorblob) [inline web/azureschandlers.go]: ?

File: web/azureschandlers.go [hunk 4]
```diff
@@ -52,22 +61,81 @@ func (ac *AppContext) loginOAuthAzureSC(w http.ResponseWriter, r *http.Request)
...
+	if www != "" {
+		re, _ := regexp.Compile("/[a-zA-Z0-9\\-]*\"")
+		match := re.FindString(www)
+		match = strings.TrimLeft(match, "/")
+		match = strings.TrimRight(match, "\"")
```
Comment (liorblob) [inline web/azureschandlers.go]: Add note here about what it's doing?

File: web/azureschandlers.go [hunk 5]
```diff
@@ -52,22 +61,81 @@ func (ac *AppContext) loginOAuthAzureSC(w http.ResponseWriter, r *http.Request)
...
+	}
+}
+
+func getTokens(tenant string, code string) (string, string, string) {
+	BaseUrl := fmt.Sprintf("https://login.microsoftonline.com/%s/OAuth2/token", tenant)
+	body := url.Values{
+		"client_id":     {conf.Options.AzureSC.ClientID},
+		"client_secret": {conf.Options.AzureSC.ClientSecret},
+		"grant_type":    {"authorization_code"},
+		"redirect_uri":  {"https://ec2-63-35-157-116.eu-west-1.compute.amazonaws.com/azuresc"},
```
Comment (liorblob) [inline web/azureschandlers.go]: doesn't look right

File: web/azureschandlers.go [hunk 6]
```diff
@@ -52,22 +61,81 @@ func (ac *AppContext) loginOAuthAzureSC(w http.ResponseWriter, r *http.Request)
...
+	var result map[string]interface{}
+
+	json.NewDecoder(r.Body).Decode(&result)
+
+	refreshToken := result["refresh_token"].(string)
+	accessToken := result["access_token"].(string)
+	expiresOn := result["expires_on"].(string)
+
+	if len(refreshToken) == 0 || len(accessToken) == 0 || len(expiresOn) == 0 {
+		return "", "", ""
```
Comment (liorblob) [inline web/azureschandlers.go]: We don't want to return an error on this case?

File: web/azureschandlers.go [hunk 7]
```diff
@@ -52,22 +61,81 @@ func (ac *AppContext) loginOAuthAzureSC(w http.ResponseWriter, r *http.Request)
...
+		"client_id":     {conf.Options.AzureSC.ClientID},
+		"client_secret": {conf.Options.AzureSC.ClientSecret},
+		"grant_type":    {"authorization_code"},
+		"redirect_uri":  {"https://ec2-63-35-157-116.eu-west-1.compute.amazonaws.com/azuresc"},
+		"code":          {code},
+	}
+
+	r, err := http.Post(BaseUrl, "application/x-www-form-urlencoded", bytes.NewBufferString(body.Encode()))
+	if err != nil {
+		return "", "", ""
```
Comment (liorblob) [inline web/azureschandlers.go]: We don't want to return an error on this case?

File: web/azureschandlers.go [hunk 8]
```diff
@@ -86,5 +154,16 @@ func (ac *AppContext) tokenOAuthAzureSC(w http.ResponseWriter, r *http.Request)
...
+		res := &domain.CustomerToken{Customer: customer, Tenant: tenant, Token: token.Token, Created: time.Now(), RefreshToken: raw["refresh_token"].(string), Subscription: token.Subscription}
+		saveErr := ac.r.SaveToken(res)
+		if saveErr != nil {
+			logrus.Warnf("Could not create token, customer ID: [%s]", customer)
+			WriteError(w, ErrInternalServer)
+			writeJSON(w, map[string]interface{}{"error": "Error while trying to set new refresh token, try to re authenticate"})
+			return
+		}
+	}
+	writeJSON(w, map[string]interface{}{"token": raw["access_token"], "expires": raw["expires_in"], "subscription_id": token.Subscription})
```
Comment (liorblob) [inline web/azureschandlers.go]: are you sure about this? `map[string]interface{}`?

File: web/azureschandlers.go [hunk 9]
```diff
@@ -52,22 +61,81 @@ func (ac *AppContext) loginOAuthAzureSC(w http.ResponseWriter, r *http.Request)
...
+
+	r, err := http.Post(BaseUrl, "application/x-www-form-urlencoded", bytes.NewBufferString(body.Encode()))
+	if err != nil {
+		return "", "", ""
+	}
+	defer r.Body.Close()
+
+	var result map[string]interface{}
+
+	json.NewDecoder(r.Body).Decode(&result)
```
Comment (liorblob) [inline web/azureschandlers.go]: Gonna trust you on this one...

File: web/azureschandlers.go [hunk 10]
```diff
@@ -86,5 +154,16 @@ func (ac *AppContext) tokenOAuthAzureSC(w http.ResponseWriter, r *http.Request)
...
 	json.Unmarshal(data, &raw)
-	writeJSON(w, map[string]interface{}{"token": raw["access_token"], "expires": raw["expires_in"]})
+	if raw["refresh_token"].(string) != refreshToken {
+		// If new refresh token is different from the one we got, update it
+		res := &domain.CustomerToken{Customer: customer, Tenant: tenant, Token: token.Token, Created: time.Now(), RefreshToken: raw["refresh_token"].(string), Subscription: token.Subscription}
+		saveErr := ac.r.SaveToken(res)
+		if saveErr != nil {
+			logrus.Warnf("Could not create token, customer ID: [%s]", customer)
+			WriteError(w, ErrInternalServer)
+			writeJSON(w, map[string]interface{}{"error": "Error while trying to set new refresh token, try to re authenticate"})
```
Comment (liorblob) [inline web/azureschandlers.go]: `WriteError` + `writeJSON`?

File: web/azureschandlers.go (hunk 6 above)
Comment (jochman) [reply web/azureschandlers.go]: its get checked in the function who's calling it

File: web/azureschandlers.go (hunk 9 above)
Comment (jochman) [reply web/azureschandlers.go]: Danke Mein Reviewer

**Pull request merged**